### ISSUE TITLE
UPDATEIFCOPY breaks on python interpreters with non-refcounted GC

### ISSUE THREAD

**Issue by njsmith** [numpy/numpy#7054]
Not sure whether this is something we should try to fix or not, but: [Pyston](http://pyston.org) and [PyPy](http://pypy.org) are both working on getting numpy running on their python implementations. Their interpreters use standard GC, rather than CPython's traditional refcounted GC. ([see](http://doc.pypy.org/en/latest/cpython_differences.html#differences-related-to-garbage-collection-strategies)) This [breaks numpy code that relies on UPDATEIFCOPY](https://gitter.im/dropbox/pyston?at=569cb5e5c391361d48ebb5de) -- operations can return arrays that contain the wrong data, only to have this data magically flipped to become correct at some later point when the GC runs. (NB Pyston may or may not switch back to refcounted GC; PyPy certainly won't.)

One possible solution to this (and really the only one that would let us reliably support Python implementations with non-refcounted GC) would be to remove the overly-magical UPDATEIFCOPY from numpy and replace it with something more explicit.

CC: @undingen @Daetalus @fijal

**Comment by njsmith** [comment]
(Note that in principle these problems can also arise with regular CPython, if you somehow manage to reference the magic UPDATEIFCOPY array from an object that participates in a reference cycle. Or just neglect to DECREF it at the appropriate moment, I guess :-). But I guess the answer there is "don't do that then".)

**Comment by seberg** [comment]
I would not disagree with trying remove it, or at least remove it from all of inside numpy and warning people about its issues.
I assume that UPDATEIFCOPY is (at least almost) only used for `out` arrays in the C-Api that cannot be handled easily. So calling a `DECREF_AND_UPDATEIFCOPY()` function could probably do the trick.
However, that might get ugly. But even fixing it up without UPDATEIFCOPY at all, is probably not hard, since my guess is all you need to remove the UPDATEIFCOPY stuff and then before decrefing the internal array add a check like:
```
if (out && (out != internal_out_arr)) {
   /* C-Api equivalent of */
    np.copyto(out, internal_out_arr, casting="unsafe");
}
```
for at least almost all functions that use it.

**Comment by njsmith** [comment]
Yeah, I'm pretty sure that all code that relies on UPDATEIFCOPY does *know* perfectly well which DECREF is supposed to be the magic one that triggers the copy-back, because otherwise it wouldn't work. So replacing this API with a less magical one where you do some explicit `_npy_update_if_copy(original, view_or_copy)` seems reasonable enough.

**Comment by mattip** [comment]
grepping for ``UPDATEIFCOPY`` shows it is used (obviously) in ``nditer(...[['updateifcopy'',...]], ...)`` but also in 

* item_selection.c:``choose`` (``PyArray_Choose``), ``take`` (``PyArray_TakeFrom``), ``put`` (``PyArray_PutTo``, ``PyArray_PutMask``) Only in nditer is it exposed to applevel. 

* arrayobject.c: ``PyArray_SetUpdateIfCopyBase`` which is used conditionally in ``PyArray_FromArray``

* calculation.c: ``argmax`` (``PyArray_ArgMax``), ``argmin`` (``PyArray_ArgMin``), ``clip`` (``PyArray_Clip``), ``arr_insert``

It seems more tests need to be written since only three tests fail under PyPy with the root cause of this issue: 

* ``test_choose_with_out (test_core.TestMaskedArrayFunctions)``
* ``test_nditer.test_iter_nested_iters_dtype_copy``
* ``test_nditer.test_iter_array_cast``

I will be starting a branch to add tests, and refactor the code called in the destructor into a new function along the lines of ` _npy_update_if_copy``. It seems this would need to be conditionally called just before returning for any of these functions, correct?

**Comment by seberg** [comment]
I am not surprised that tests are missing, these are easy tests to miss (out array, and then the out array needs some type of cast or has wrong strides), plus a lot of the functions have not been changed in a long time. So yeah, will need tests for all the modified code paths....

Yes, it should be called just before returning. There must be a DECREF somewhere which specifically causes the UPDATEIFCOPY operation, so you will have to add some other logic right where that decref is.

**Comment by mattip** [comment]
An update - I have made progress in the pull request #9269. I refactored all the places where ``Py_DECREF`` is used to trigger the writeback, and have replaced them with a new API function ``PyArray_ResolveUpdateIfCopy``. 

I propose to turn ``nditers`` into
context managers, so that if necessary the magic happens on the ``__exit__``. This model seems to work well, and is implemented in the pull request. 

It seems I also need to do the same thing to ``flatiters``, which suffer from exactly the same issue/feature. Does this approach make sense? Is code like ``x.flat.__array__()`` used in the wild?
Note that ``y = np.array(x.flat)`` does not suffer from ``UPDATEIFCOPY`` since a new array is returned, changing y will not change x. Deprecating the ``__array__`` method on flatiter would also solve the problem

**Comment by njsmith** [comment]
It's very weird that `.flat.__array__()` uses UPDATEIFCOPY. There's some discussion in #9244. Honestly I think it should just stop doing that and return a regular copy; it's not at all consistent with how `__array__` is used in other cases, and #9249 made it so that `__array__`'s main user detects this case and forces a copy anyway, so the only code that would even notice is code that's doing some dicey things with numpy's internal APIs.

@charris, any opinion?

**Comment by njsmith** [comment]
It might make sense to ignore the `.flat` issue in #9269 in any case; no need to fix everything in one PR.

**Comment by charris** [comment]
@njsmith I think it should return a copy. How to get there might be dicey. If we issue a `FutureWarning` we will spam users who probably don't care, and I expect that is almost all of them. The easiest thing might be to just do it, make a release note, and see if anyone notices.

**Comment by charris** [comment]
@njsmith Note that #9249 was not merged and that in the case of an already flat contiguous array `*.flat.__array__()` returns a view, which is makes the `UPDATEIFCOPY` case at least consistent.

Does nditer expose  `UPDATEIFCOPY` in a similar way?

**Comment by njsmith** [comment]
@charris: Hmm. Do we use `__array__` internally, e.g. for implementing `__getitem__` or `__setitem__` on `.flat`?

Skimming some downstream projects with `rg -w \\.flat`:

Looking at scikit-learn I see lots of code like `a.flat[::N] += 1`. (There's also some `a.flat[someslice] = something`, `for k in a.flat: ...`, `assert_array_equal(a.flat, ...)`, `a.flat[0]`, and a `a.flat[something] = b.float[something]`. Mostly they seem to be using it as an idiom for writing to the diagonal (`mat.flat[::mat.shape[0] + 1] = ...`).

matplotlib uses `a.flat` or `a.flat[1:]` in iterator expressions a lot (`for x in a.flat:`, `zip(a.flat, b.flat)`, etc.), but I don't think this involves `__array__`. They also have a few places where they do stuff like `column_stack((a.flat, b.flat))`, which probably does invoke `__array__`, but doesn't try writing to the resulting array, so that's fine.

scikit-image has code like `arr[label.flat == bg_label]`, `np.argsort(a.flat)` `np.histogram(image.flat)`, `np.interp(image.flat, ...)`, more iteration, more stacking, .... so again it definitely does invoke `__array__` implicitly but never seems to write to the result, just pass it into some other operation.

scipy does `a.flat = np.nan`, plus some of the stuff already mentioned above.

It looks tentatively like we could get away with having `flat.__array__` switch to returning a copy, either unconditionally, or just in cases where it can't return a view.

That would make this a similar transition to the `np.diagonal` -- we could reuse `NPY_ARRAY_WARN_ON_WRITE` to issue deprecation warnings, and/or make it return a read-only array for the transition period (in cases when it used to return a view or UPDATEIFCOPY array and in the future will return a copy).

**Comment by njsmith** [comment]
Oh right, and I meant to say -- even if we do use it internally ourselves, that's not a problem; for our internal uses we could have a private internal version of `__array__` uses UPDATEIFCOPY_FIXED, and make sure we resolve the copy properly.

**Comment by mattip** [comment]
Last comment first: internally we can force the writeback as I did in commit 9f15a2e4c6

About user facing code with the ``__array__`` method call: I would prefer to force use of a context manager if the call to ``__array__`` forces creation of a ndarray with the ``UPDATEIFCOPY`` flag and the user tries to write to the view, rather than maybe silently creating a copy. If I were a naive user, code which works this way today would be very confusing
```
x = np.arange(9).reshape(3,3)
a = x.flat.__array__()
# assign to the view
a[:] = 0
assert np.sum(x) == 0

a = x.T.flat.__array__()
a[:] = 1
# HUH ?????????
assert np.sum(x) == 0
# trigger resolution - if it is a copy there is no way to write back to the original
del a
assert np.sum(x) == 9
```

As a user, I would prefer that code trigger an error telling me to use a context manager::
```
a = x.T.flat.__array()
a[:] = 1  # <----- error triggered here "use a context manager when UPDATEIFCOPY flag is True"

with x.T.flat.__array__() as a:
    a[:] = 10
assert np.sum(x) ==  90 
```

**Comment by charris** [comment]
> would prefer to force use of a context manager if the call to __array__ forces creation of a ndarray with the UPDATEIFCOPY flag and the user tries to write to the view,

Seeing as how the user currently doesn't know if they have a view or a copy without checking the flags, I don't see the point of being fancy. In fact, I would regard writing to the result to be a bug in general because no one knows where it ends up.

**Comment by njsmith** [comment]
I agree with @charris that this is gross and totally unlike every other `__array__` method and we don't want to keep it this way; the discussion should be about how to stop supporting this, not about about some context manager wackiness to keep it working. (@mattip: note that it's *very* rare to call `__array__`  manually -- it's normally called implicitly by the array coercion machinery that gets invoked every time you pass an object to any numpy function. So it's being called by things like `np.vstack((a.flat, b.flat))` or `np.arange(10) + a.flat` or whatever.)

@charris: I guess right now `a.flat.__array__` does always return a view or an UPDATEIFCOPY pseudo-view, so right now code like `np.array(a.flat)[0] = 1` does work reliably regardless of `a`'s layout. (Of course this is fragile b/c it's only reliable if they immediately drop their reference to the returned array, but nonetheless it is possible that there is existing code out there that works correctly and relies on this.)

So I guess our options are:

* make `a.flat.__array__` start returning the same array it does now but with the `NPY_WARN_ON_WRITE` flag set, then eventually switch it to returning a read-only copy (so any code that expects writing to it to work will get a clean error and hopefully get fixed), then eventually switch it to returning a writable copy.

* immediately switch it to returning a read-only copy, then eventually switch it to returning a writable copy

* immediately switch it to returning a writable copy

The last one makes me nervous but given how unusual this case is and that we'd like to enable pypy support like, this year rather than next year if possible, I'd be OK with going straight to the read-only return ("option 2").

**Comment by charris** [comment]
@njsmith 2 looks like a reasonable compromise.

**Comment by charris** [comment]
I'll put up a PR.

**Comment by charris** [comment]
Note the following inconvenience

>`class.__array__([dtype])`
>
>    If a class (ndarray subclass or not) having the `__array__` method is used as the output object of an ufunc, results will be written to the object returned by `__array__`. Similar conversion is done on input arrays.

I think this means that `a.flat` should not be used as an output array ;)

**Comment by charris** [comment]
See #9447. I left the views of contiguous arrays alone.

**Comment by mattip** [comment]
@charris did 6de4bf1 ever get merged to upstream? I am not seeing it on numpy/numpy/master

**Comment by charris** [comment]
@mattip Nope, I was waiting for someone else to review and do the honors.

**Comment by charris** [comment]
It is better to refer to the PR, #9447.

**Comment by njsmith** [comment]
@mattip: I just merged #9447.

**Comment by mattip** [comment]
since pull request #9447 was merged, this mega-issue is now waiting for pull request #9639 and resolution of issue #9714

**Comment by mattip** [comment]
pull request #9639 was merged, I started pull request #9998 to resolve issue #9714.

**Comment by mattip** [comment]
Closed via PR #9998
